data_structures/src/mainnet_validations.rs

```diff
@@ -16,6 +18,143 @@ pub const FIRST_HARD_FORK: Epoch = 192000;
+                        if (bit_counter.votes * 100) / bit_counter.period >= 80 {
+                            // An offset of 21 is added to ensure that the activation of the WIP is
+                            // achieved with consolidated blocks
+                            self.wip_activation
+                                .insert(bit_counter.wip.clone(), epoch + 21);
+                        }
+                        bit_counter.votes = 0;
+                    }
+                }
+            }
+        }
+        self.bit_tapi_counter.last_epoch = epoch;
+    }
+
+    pub fn initialize_wip_information(&mut self) -> (Epoch, HashSet<String>) {
+        // Hardcoded information about WIPs
+        let mut bits = vec![vec![]; 32];
+        let wip_0014 = BitVotesCounter {
+            votes: 0,
+            period: 26880,
+            wip: "WIP0014".to_string(),
+            init: 500000,
+            end: 850000,
+        };
+        bits[0].push(wip_0014);
+
+        // Assessment of new WIPs
+        let mut min_epoch = u32::MAX;
+        let mut old_wips = HashSet::default();
+        for (i, wips) in bits.into_iter().enumerate() {
+            for wip in wips {
+                if self.bit_tapi_counter.contains(i, &wip.wip) {
+                    old_wips.insert(wip.wip);
+                } else {
+                    if wip.init < min_epoch {
+                        min_epoch = wip.init;
+                    }
+                    self.bit_tapi_counter.insert(i, wip);
+                }
+            }
+        }
+
+        (min_epoch, old_wips)
+    }
+}
+
+/// Struct that count the positives votes of a WIP
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct BitVotesCounter {
+    pub votes: u32,
+    pub period: Epoch,
+    pub wip: String,
+    pub init: Epoch,
+    pub end: Epoch,
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct BitTapiCounter {
+    pub info: [Vec<BitVotesCounter>; 32],
+    pub last_epoch: Epoch,
+}
+
+impl BitTapiCounter {
+    pub fn get_mut(&mut self, bit: usize, epoch: &u32) -> Option<&mut BitVotesCounter> {
+        match self.info.get_mut(bit) {
+            Some(bit_info) => {
+                for i in bit_info {
+                    if *epoch >= i.init && *epoch < i.end {
+                        return Some(i);
+                    }
+                }
+
+                None
+            }
+            None => None,
+        }
+    }
+
+    pub fn insert(&mut self, k: usize, v: BitVotesCounter) {
+        match self.info.get_mut(k) {
+            Some(bit_info) => {
+                bit_info.push(v);
+            }
+            None => {
+                self.info[k] = vec![v];
+            }
+        }
+    }
+
+    pub fn contains(&self, bit: usize, wip: &str) -> bool {
+        match self.info.get(bit) {
+            Some(bit_info) => {
+                for i in bit_info {
+                    if i.wip.eq(wip) {
```

Comment:
```suggestion
                    if i.wip == wip {
```

Reply:
No, wait, this for loop is `.any()`

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any